### ISSUE TITLE
[bug]: `no-direct-set-state-in-use-(layout?)-effect` should warn only for the `set` function

### ISSUE THREAD

**Issue by AhmedBaset** [Rel1cx/eslint-react#665]
### Describe the bug

Given a code like that: 
```js
const [fn] = useState(() => () => "Function");
// ...
useEffect(() => {
  fn();
}, [...])
```

If the state is a function, invoking it will result in an error because we do not check which value is being called.

https://github.com/Rel1cx/eslint-react/blob/c4f84dabcd7075fe46ee719c78736d33b23800c2/packages/plugins/eslint-plugin-react-hooks-extra/src/rules/no-direct-set-state-in-use-effect.ts#L65-L69

`O.some()` will match both `data` and `setData`

### Reproduction

```js
const [fn] = useState(() => () => "Function");
// ...
useEffect(() => {
  fn();
}, [...])
```

### Expected behavior

It should only warn for calling `setState`

### Platform and versions

```Text
@eslint-react: ^1.5.27
```

**Comment by Rel1cx** [comment]
Yes! Good Catch. Thanks!

**Comment by AhmedBaset** [comment]
Thanks for addressing

I'm curious about the duplication between `no-direct-set-state-in-use-effect` and `no-direct-set-state-in-use-layout-effect`. Why they are not a single rule or at least single functionality.

**Comment by Rel1cx** [comment]
> Thanks for addressing
> 
> I'm curious about the duplication between `no-direct-set-state-in-use-effect` and `no-direct-set-state-in-use-layout-effect`. Why they are not a single rule or at least shared functionality?

There are multiple rules because it gives the user the ability to control whether or not this rule applies to other terms just by switching the ESLint rules on or off, rather than having to add options or exceptions to the single rule to accomplish this.

It also prevents any potential differences between `useEffect` and `useLayoutEffect` in future versions of React that might change the behavior of a single rule.

And if other terms such as `useLayoutEffect` are deprecated, we just deprecate the `no-direct-set-state-in-use-layout-effect` rule, and the `no-direct-set-state-in-use-effect` rule doesn't get affected.

Although these benefits are certainly not obvious in the case of `no-direct-set-state-in-use-(layout?)-effect`, and they are more obvious with some of the other rules in the project such as `no-children-*` or `no-set-state-in-component-did-*`.

> or at least shared functionality

Agreed, I would extract the shared functionality in one place to avoid modifying the same code in multiple places.
